WalletWasabi/BitcoinCore/Configuration/CoreConfig.cs

```diff
@@ -0,0 +1,131 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using WalletWasabi.Helpers;
+
+namespace WalletWasabi.BitcoinCore.Configuration
+{
+	public class CoreConfig
+	{
+		private List<CoreConfigLine> Lines { get; }
+
+		public int Count => Lines.Count;
+
+		public CoreConfig(CoreConfig coreConfig)
+		{
+			Lines = coreConfig.Lines.ToList();
+		}
+
+		public CoreConfig()
+		{
+			Lines = new List<CoreConfigLine>();
+		}
+
+		public IDictionary<string, string> ToDictionary()
+		{
+			var myDic = new Dictionary<string, string>();
+			foreach (var line in Lines)
+			{
+				if (line.HasKeyValuePair)
+				{
+					myDic.Add(line.Key, line.Value);
+				}
+			}
+
+			return myDic;
+		}
+
+		public override string ToString() => $"{Guard.Correct(string.Join(Environment.NewLine, Lines))}{Environment.NewLine}"; // Good practice to end with a newline.
+
+		/// <summary>
+		/// TryAdd and AddOrUpdate are sisters. TryAdd always considers the first occurrence of a key as valid, while AddOrUpdate considers the last occurrence of it.
+		/// </summary>
+		public void TryAdd(string configString)
+		{
+			foreach (var line in ParseConfigLines(configString))
+			{
+				if (line.HasKeyValuePair)
+				{
+					var foundLine = Lines.FirstOrDefault(x => x.Key == line.Key);
+					if (foundLine is null)
+					{
+						Lines.Add(line);
+					}
+				}
+				else
+				{
+					Lines.Add(line);
+				}
+				RemoveFirstEmptyDuplication();
+			}
+		}
+
+		/// <summary>
+		/// TryAdd and AddOrUpdate are sisters. TryAdd always considers the first occurrence of a key as valid, while AddOrUpdate considers the last occurrence of it.
+		/// </summary>
+		public bool AddOrUpdate(string configString)
+		{
+			var ret = false;
+			foreach (var line in ParseConfigLines(configString))
+			{
+				if (line.HasKeyValuePair)
+				{
+					var foundLine = Lines.FirstOrDefault(x => x.Key == line.Key);
+					if (foundLine is null)
+					{
+						Lines.Add(line);
+						ret = true;
+					}
+					else if (foundLine.Value != line.Value)
+					{
+						Lines.Remove(foundLine);
+						Lines.Add(line);
+						ret = true;
+					}
+				}
+				else
+				{
+					Lines.Add(line);
+					ret = true;
+				}
+				ret = ret || RemoveFirstEmptyDuplication();
+			}
+
+			return ret;
+		}
+
+		private bool RemoveFirstEmptyDuplication()
+		{
+			CoreConfigLine toRemove = null;
+			for (int i = 1; i < Lines.Count; i++)
+			{
+				CoreConfigLine currentLine = Lines[i];
+				CoreConfigLine prevLine = Lines[i - 1];
+				if (string.IsNullOrWhiteSpace(currentLine.Line) && string.IsNullOrWhiteSpace(prevLine.Line))
+				{
+					toRemove = currentLine;
+					break;
+				}
+			}
+
+			if (toRemove != null)
+			{
+				Lines.Remove(toRemove);
+				return true;
+			}
+
+			return false;
```

Comment:
Same for this one.
```
for (int i = 1; i < Lines.Count; i++)
{
	CoreConfigLine currentLine = Lines[i];
	CoreConfigLine prevLine = Lines[i - 1];
	if (string.IsNullOrWhiteSpace(currentLine.Line) && string.IsNullOrWhiteSpace(prevLine.Line))
	{
		Lines.Remove(currentLine);
		return true;
	}
}

return false;
```

Also shouldn't we declare `CoreConfigLine currentLine` & `CoreConfigLine prevLine` outside the for loop to net keep creating instances of them each time?

Reply:
@lontivero It's a list, not a string. @yahiheb Only concurrent collections are safe to be modified while they're being iterated through.

Reply:
@nopara73 I think the code I suggested doesn't change anything. Can you check again.